### ISSUE TITLE
NLP Features

### ISSUE THREAD

**Issue by seanzhou1207** [current12/Stat-222-Project#20]
- [x] Sentiment analysis: FINBERT
          ![Count Positive + 1](https://github.com/current12/Stat-222-Project/assets/74280000/a802a384-f6dc-4df2-b1fe-6d79a0cb8efd)
[source](https://deliverypdf.ssrn.com/delivery.php?ID=299118127070009118110116069108121107103081020021005061018087067068096123022117001030107045042043006097108088118000113064026119042090023016077067107090113119066120074027079048119082006014029100086004097117100126027005111100086084078008008069116090100070&EXT=pdf&INDEX=TRUE)

- [x] Sentiment analysis (dictionary) - formula above but using positive and negative word counts from Harvard dictionary

- [ ] Store average FINBERT embeddings for each call as well, may use them

- [x] Overall tone: [source](https://www.sciencedirect.com/science/article/pii/S0378426611002901?casa_token=IspUwTIIcMgAAAAA:4x8VDW1hDznHg0AXs32ViKMxgQFp-nZs1cRyQe_m-YnbILXt-JNfyhHIMGdMIZizk80AzoP4)
![Pasted Graphic](https://github.com/current12/Stat-222-Project/assets/74280000/2b76568e-8e24-4117-a7f0-e4cc74c0f8df)

- [x] Numeric transparency: ratio of numbers to words [source](http://(https//www.spglobal.com/marketintelligence/en/news-insights/blog/analyzing-sentiment-in-quarterly-earnings-calls-q2-2022))
- [x] Engagement - Number of Question Marks
- [x] Gunning-Fog Readability
- [x] Call length in words. EDA evidence shows this is likely to be a good predictor. Please adapt the code below.
```
# Tokenize the text
df['transcript_tokens'] = df['transcript'].apply(word_tokenize)
# Words: strip punctuation and make lowercase
# Regex for not being solely punctuation
df['transcript_words'] = df['transcript_tokens'].apply(lambda x: [token for token in x if re.search(r'^[^\w]+$', token) == None])
```

**Comment by ijyliu** [comment]
For embeddings, I'd suggest finbert and also possibly a specific transformer classifier finetuned for this task

First link on positive v. negative words doesn't work. Sounds pretty straightforward to me though. I'd say you can go ahead and download that dictionary because that's a baseline we will definitely use.

How do you plan to use Loughran McDonald finance dictionary? What's it's value add in addition to just the positive/negative words? How easy is that to download and use?

What is Alexandria?

For tone, what dictionaries would you plan to use to get values of active/passive, etc for the tone PCA?

I think for analyst engagement we may be limited to just counting question marks overall because it will be hard to parse out the questions segment.

Have you looked into the textstat package yet? I think that will help with feature 4. There are also other easy text features it's capable of generating easily.

I'd possibly suggest trimming the number of dictionaries and simple features in favor of spending more time on neural approaches. It seems like the sentiment positivity feature is kind of subsumed in the tone feature, so I would pick just one of those. I'd bet the paper for the tone feature has a comparison with just the ratio of positive/negative words vs. throwing it plus other stuff into PCA. Is the performance gain from PCA vs. just that ratio large? PCA would be less interpretable and require more work and dictionaries.

**Comment by ijyliu** [comment]
@OwenLin2001 @current12 what do you think? we can also send to libor once we've refined

**Comment by ijyliu** [comment]
![image](https://github.com/current12/Stat-222-Project/assets/67396209/3e8e236a-d718-4102-a236-f0237278247f)

**Comment by ijyliu** [comment]
@seanzhou1207 

I'd get started coding these (starting with the dictionaries and simple ones) if you haven't already

You can make a new notebook and test/debug on all_data_sample.csv and then scale up to all_data.parquet

Then you can save as all_data_with_nlp.parquet or something

**Comment by ijyliu** [comment]
input file is now `all_data_fixed_quarter_dates.parquet`

please output `all_data_fixed_quarter_dates_NLP.parquet` or something (parquet load time and upload/download speeds are much faster, storage space is less than half as much (my computer is filling up 💀 + we might be approaching Box limits), and it will also be super easy to load just the feature columns and not the call transcript itself in future)

**Comment by ijyliu** [comment]
parallelizing feature construction:

this is an embarassingly parallel situation - each row of the data/each transcript is entirely independent. so we have lots of options

- Continue to work locally and use dask, modin, other python libraries for multi-core. Split features into a separate script/notebook so they can be run independently - try to separate fast from slow code
- SCF. You can apply for an individual account and they will give you access quickly. it's then simple to parallelize using the SLURM scheduler (I have experience with this). The only real pain point is that we will have to find a way to move data to/from Box and SCF. Several options here, such as the Box API (ask Owen? you could use the api to read/write the df to Box from within your script I think) which we've already been using, or rclone/rsync (a shell script to save the data to your home directory or a shared project directory and keep it up to date). it might also take a little time to read the SCF documentation on conda environments, thankfully we have environment.yml in this repo already, so that will make it easy to build
- AWS/various other cloud compute platforms - this is actually more bureaucratic and difficult than I envisioned and hence not recommended... we can still use them for autoML and/or actual training, but don't suggest getting into it for feature construction

**Comment by ijyliu** [comment]
hey @seanzhou1207 do you anticipate having missing values for any of these constructed NLP variables - like if the ratio has zero in the denominator?

**Comment by seanzhou1207** [comment]
> hey @seanzhou1207 do you anticipate having missing values for any of these constructed NLP variables - like if the ratio has zero in the denominator?

yes, but less than 5 missing values out of all 7000

**Comment by ijyliu** [comment]
i found this package for finbert: https://pypi.org/project/finbert-embedding/

of note: finbert gives sentence level embeddings, but it's been proposed that these aren't super meaningful. anyway, there are several ways to think about aggregating finbert representations to a measure of sentiment. we could take an average embedding for the entire call. or we could calculate sentiment for each sentence using the embeddings, then aggregate this into sentiment for each call. i'd suggest calculating sentiment for each sentence/other granular unit, then aggregating that into a measure of call sentiment.

**Comment by ijyliu** [comment]
Use: https://pypi.org/project/finbert-embedding/

**Comment by seanzhou1207** [comment]
- Visualization for the NLP features
- Use Harvard IV-4 Psychosocial to get the other tones
- Use finbert embedding to get positivity score
- Count question marks to get analyst engagement 
- Combine all NLP features into 1 parquet file

**Comment by ijyliu** [comment]
Added "Store average FINBERT embeddings for each call as well, may use them" as a feature

**Comment by seanzhou1207** [comment]
<img width="116" alt="image" src="https://github.com/current12/Stat-222-Project/assets/74280000/90723ccd-a59c-44c3-a2ea-0cef43daf6cb">
So far finished all features except for the "tone". Issue: all words from the Harvard dictionary are non-stemmed and some have multiple meanings. It's hard to map the words in earning call to the dictionary.

**Comment by ijyliu** [comment]
just do the best you can. if we do eda and discover it's not helpful we can just ignore it

**Comment by ijyliu** [comment]
@seanzhou1207 did you save finbert embeddings or average finbert embeddings for each call? not seeing them

i can go make them. they seem like they might be directly useful

**Comment by ijyliu** [comment]
@seanzhou1207 what is the difference between column `readability` and column `gf_score`?

**Comment by ijyliu** [comment]
@seanzhou1207 

what is the difference between these two variables

![image](https://github.com/current12/Stat-222-Project/assets/67396209/13b49ea3-b154-406b-a82e-db5458d88c59)

**Comment by seanzhou1207** [comment]
> @seanzhou1207 did you save finbert embeddings or average finbert embeddings for each call? not seeing them
> 
> i can go make them. they seem like they might be directly useful

No, FINEBRT was taking too long so I ended up getting positive and negative word counts using the Harvard dictionary as well.

**Comment by seanzhou1207** [comment]
> @seanzhou1207
> 
> what is the difference between these two variables
> 
> ![image](https://private-user-images.githubusercontent.com/67396209/318611124-13b49ea3-b154-406b-a82e-db5458d88c59.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwMzcwNzAsIm5iZiI6MTcxMjAzNjc3MCwicGF0aCI6Ii82NzM5NjIwOS8zMTg2MTExMjQtMTNiNDllYTMtYjE1NC00MDZiLWE4MmUtZGI1NDU4ZDg4YzU5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAyVDA1NDYxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU1MzU0ZDdlNDM0NDFmM2FjNGI5ZTMxN2FhNDVkNWYyMzBiYjM2ZGUyYjc1MWY5MDNiYjM3MDI3MTMzZTQyYzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PWGfDLyvjKxOFLXOBaWkopAU4JyLwj3rHb4YjlHbL74)

No difference. sorry

**Comment by ijyliu** [comment]
I think finbert is pretty critical. For autogluon at least the sentiment
doesn't perform super good. And I think the profs expect us to use it since
that was pretty key to the project topic.

Just do it in a `.py` file and request an A100. I'll go get an sbatch for you
to use for it. If you can't figure out how to do it on GPU with the package
we can go get it from huggingface

**Comment by ijyliu** [comment]
heres a shell script

https://github.com/current12/Stat-222-Project/blob/main/Code/Exploratory%20Data%20Analysis/All%20Data/EDA_NER.sh

**Comment by ijyliu** [comment]
> > @seanzhou1207
> > what is the difference between these two variables
> > ![image](https://private-user-images.githubusercontent.com/67396209/318611124-13b49ea3-b154-406b-a82e-db5458d88c59.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwMzcwNzAsIm5iZiI6MTcxMjAzNjc3MCwicGF0aCI6Ii82NzM5NjIwOS8zMTg2MTExMjQtMTNiNDllYTMtYjE1NC00MDZiLWE4MmUtZGI1NDU4ZDg4YzU5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAyVDA1NDYxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU1MzU0ZDdlNDM0NDFmM2FjNGI5ZTMxN2FhNDVkNWYyMzBiYjM2ZGUyYjc1MWY5MDNiYjM3MDI3MTMzZTQyYzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PWGfDLyvjKxOFLXOBaWkopAU4JyLwj3rHb4YjlHbL74)
> 
> No difference. sorry

ok, i'm going to delete `readability` then

**Comment by ijyliu** [comment]
continued finbert construction:

- split into more separate programs + do smaller numbers of calls at a time
- make sure you have a separate log for each split/program so you will know if any specific one is failing
- spend a little time trying to see if more dask internals can be printed out (like progress, etc.)

**Comment by ijyliu** [comment]
finbert sentiment analysis complete, average finbert embeddings moved to issue #81
